app/controllers/image_proxy_controller.rb

```diff
@@ -0,0 +1,35 @@
+# frozen_string_literal: true
+
+##
+# Controller for proxying HTTP images
+class ImageProxyController < ApplicationController
+  before_action :validate_token
+
+  def access
+    send_data proxied_response.body, type: proxied_response.content_type, disposition: 'inline'
+  end
+
+  private
+
+  ##
+  # To prevent outside applications from using this proxy
+  def validate_token
+    render plain: 'invalid token', status: :forbidden unless valid_authenticity_token?(session, image_token)
+  end
+
+  def proxied_response
+    @proxied_response ||= begin
+      benchmark "Fetch #{image_url}" do
```

Comment:
Should we look into Rails.cache here?

Reply:
If we throw a rails cache around this would that assuage some of the concern about authenticity tokens breaking browser cache?

Reply:
I've popped a commit on there that the ability to do this. One drawback is that I don't think we can stream and cache the responses at the same time :(

Reply:
Ok.. Just added one more.. Using a referer check instead of token based? This will hopefully enable browser caching and still be a useful deterrent, though of course it can be spoofed.